torch_geometric/explain/algorithm/subgraphX_explainer.py

```diff
@@ -0,0 +1,646 @@
+        high2low (:obj:`bool`): Whether to expand children tree node
+            from high degree nodes to low degree nodes.
+        node_idx (:obj:`int`): The target node index to extract
+            the neighborhood.
+        score_func (:obj:`Callable`): The reward function for
+            tree node, such as mc_shapely and mc_l_shapely.
+    """
+    def __init__(
+        self,
+        X: torch.Tensor,
+        edge_index: torch.Tensor,
+        num_hops: int,
+        n_rollout: int = 10,
+        min_atoms: int = 3,
+        c_puct: float = 10.0,
+        expand_atoms: int = 14,
+        high2low: bool = False,
+        node_idx: int = None,
+        score_func: Callable = None,
+        device="cpu",
+    ):
+
+        self.X = X
+        self.edge_index = edge_index
+        self.device = device
+        self.num_hops = num_hops
+        self.data = Data(x=self.X, edge_index=self.edge_index)
+        graph_data = Data(x=self.X,
+                          edge_index=remove_self_loops(self.edge_index)[0])
+        self.graph = to_networkx(graph_data, to_undirected=True)
+        self.data = Batch.from_data_list([self.data])
+        self.num_nodes = self.graph.number_of_nodes()
+        self.score_func = score_func
+        self.n_rollout = n_rollout
+        self.min_atoms = min_atoms
+        self.c_puct = c_puct
+        self.expand_atoms = expand_atoms
+        self.high2low = high2low
+        self.new_node_idx = None
+
+        # extract the sub-graph and change the node indices.
+        if node_idx is not None:
+            if isinstance(node_idx, torch.Tensor):
+                node_idx = node_idx.item()
+
+            self.ori_node_idx = node_idx
+            self.ori_graph = copy.copy(self.graph)
+            x, edge_index, subset, edge_mask, kwargs = self.__subgraph__(
+                node_idx, self.X, self.edge_index, self.num_hops)
+            self.data = Batch.from_data_list(
+                [Data(x=x, edge_index=edge_index)])
+            self.graph = self.ori_graph.subgraph(subset.tolist())
+            mapping = {int(v): k for k, v in enumerate(subset)}
+            self.graph = nx.relabel_nodes(self.graph, mapping)
+            self.new_node_idx = torch.where(
+                subset == self.ori_node_idx)[0].item()
+            self.num_nodes = self.graph.number_of_nodes()
+            self.subset = subset
+
+        self.root_coalition = sorted([node for node in range(self.num_nodes)])
+        self.MCTSNodeClass = partial(
+            MCTSNode,
+            data=self.data,
+            ori_graph=self.graph,
+            c_puct=self.c_puct,
+            device=self.device,
+        )
+        self.root = self.MCTSNodeClass(self.root_coalition)
+        self.state_map = {str(self.root.coalition): self.root}
+
+    def set_score_func(self, score_func):
+        self.score_func = score_func
+
+    @staticmethod
+    def __subgraph__(node_idx, x, edge_index, num_hops, **kwargs):
```

Comment:
I haven't checked but this seems like it should either already exist or would be nice to abstract out of this specific class.